src/secrets/src/lib.rs

```diff
@@ -58,25 +63,48 @@ pub struct SecretsReaderConfig {
 /// Securely reads secrets that are managed by a [`SecretsController`].
 ///
 /// Does not provide access to create, update, or delete the secrets within.
+#[derive(Debug, Clone)]
 pub struct SecretsReader {
-    config: SecretsReaderConfig,
+    config: Arc<SecretsReaderConfig>,
 }
 
 impl SecretsReader {
     pub fn new(config: SecretsReaderConfig) -> Self {
-        Self { config }
+        Self {
+            config: Arc::new(config),
+        }
     }
 
     /// Returns the contents of a secret identified by GlobalId
+    ///
+    /// This `read` will return a complete version of the secret even though `SecretsReader` is `Clone` and we could
+    /// have concurrent reads.  It will not return, e.g., one block from v1 and another from v2.
+    ///
+    /// - On Linux / OSX filesystems, `File::open` will hold a handle open so that even if the file is deleted, we still
+    ///   continue to read from it.  This means a SecretOp::Delete followed by a SecretOp::Ensure can never "swap out"
+    ///   data mid-`read_to_end`.
+    /// - We do not allow editing secrets which _would_ expose us to this issue.
+    ///
+    /// (N.B. Were we ever to run with Windows / NTFS, this would also work properly _and_ mid-read edits would be
+    /// disallowed)
     pub fn read(&self, id: GlobalId) -> Result<Vec<u8>, anyhow::Error> {
-        let file_path = self.config.mount_path.join(id.to_string());
-        Ok(fs::read(file_path)?)
+        let file_path = self.config.mount_path.join(id.borrow().to_string());
+
+        // Inlined the std::fs::read impl because correctness requires and impl that holds the same `File` handle open
+        let mut file = File::open(file_path)?;
+        let mut buf = Vec::new();
+        file.read_to_end(&mut buf)?;
+        Ok(buf)
+    }
+
+    pub fn read_string(&self, id: GlobalId) -> anyhow::Result<String> {
+        String::from_utf8(self.read(id)?).context("converting secret value to string")
     }
 
     /// Returns the path of the secret consisting of a configured base path
     /// and the GlobalId
     pub fn canonical_path(&self, id: GlobalId) -> Result<PathBuf, anyhow::Error> {
-        let path = self.config.mount_path.join(id.to_string());
+        let path = self.config.mount_path.join(id.borrow().to_string());
```

Comment:
Spurious?

Reply:
whoops -- was necessary in a past version but not anymore.  will remove here and above